lib/ClientSai.cpp

```diff
@@ -999,6 +989,149 @@ sai_status_t ClientSai::waitForGetStatsResponse(
     return status;
 }
 
+sai_status_t ClientSai::queryStatsCapability(
+        _In_ sai_object_id_t switchId,
+        _In_ sai_object_type_t objectType,
+        _Inout_ sai_stat_capability_list_t *stats_capability)
+{
+    MUTEX();
+    SWSS_LOG_ENTER();
+    REDIS_CHECK_API_INITIALIZED();
+
+    if (stats_capability && stats_capability->list)
+    {
+        // clear input list, since we use serialize to transfer values
+        for (uint32_t idx = 0; idx < stats_capability->count; idx++)
+	{
+            stats_capability->list[idx].stat_enum = 0;
+            stats_capability->list[idx].stat_modes = 0;
+	}
+    }
+
+    auto switchIdStr = sai_serialize_object_id(switchId);
+    auto objectTypeStr = sai_serialize_object_type(objectType);
+    const std::string list_size = std::to_string(stats_capability->count);
+
+    const std::vector<swss::FieldValueTuple> entry =
+    {
+        swss::FieldValueTuple("OBJECT_TYPE", objectTypeStr),
+	swss::FieldValueTuple("LIST_SIZE", list_size)
+    };
+
+    SWSS_LOG_DEBUG(
+            "Query arguments: switch %s, object type: %s, count: %s",
+            switchIdStr.c_str(),
+            objectTypeStr.c_str(),
+            list_size.c_str()
+    );
+
+    // This query will not put any data into the ASIC view, just into the
+    // message queue
+
+    m_communicationChannel->set(switchIdStr, entry, REDIS_ASIC_STATE_COMMAND_STATS_CAPABILITY_QUERY);
+
+    auto status = waitForQueryStatsCapabilityResponse(stats_capability);
+
+    return status;
+}
+
+sai_status_t ClientSai::waitForQueryStatsCapabilityResponse(
+        _Inout_ sai_stat_capability_list_t* stats_capability)
+{
+    SWSS_LOG_ENTER();
+
+    swss::KeyOpFieldsValuesTuple kco;
+
+    auto status = m_communicationChannel->wait(REDIS_ASIC_STATE_COMMAND_STATS_CAPABILITY_RESPONSE, kco);
+
+    if (status == SAI_STATUS_SUCCESS)
+    {
+        const std::vector<swss::FieldValueTuple> &values = kfvFieldsValues(kco);
+
+        if (values.size() != 3)
+        {
+            SWSS_LOG_ERROR("Invalid response from syncd: expected 3 value, received %zu", values.size());
+
+            return SAI_STATUS_FAILURE;
+        }
+
+        const std::string &stat_enum_str = fvValue(values[0]);
+        const std::string &stat_modes_str = fvValue(values[1]);
+        const uint32_t num_capabilities = std::stoi(fvValue(values[2]));
+
+        SWSS_LOG_DEBUG("Received payload: stat_enums = '%s', stat_modes = '%s', count = %d",
+			stat_enum_str.c_str(), stat_modes_str.c_str(), num_capabilities);
+
+        stats_capability->count = num_capabilities;
+
+        size_t stat_enum_position = 0;
+        size_t stat_modes_position = 0;
+        for (uint32_t i = 0; i < num_capabilities; i++)
+        {
+		/* Populate stat_enum */
+            size_t old_stat_enum_position = stat_enum_position;
+            stat_enum_position = stat_enum_str.find(",", old_stat_enum_position);
```

Comment:
Is it possible to leverage `tokenize` which returns all the substring split by a splitter, eg `,`?

Reply:
Tokenize should also work fine. I have opted this, as this algorithm is used in various Attribute/enum processing APIs (including waitForQueryAattributeEnumValuesCapabilityResponse). I would stick to current way of processing the Attributes/Enums.

Reply:
Hi @rpmarvell 
could you please clarify what optimization in your implementation compared to using `tokenize`?
In case the current way is better, I would like to suggest implementing a common function (maybe in the base class) and calling it in `ClientSai`, `RedisRemoteSaiInterface` instead of duplicating it in two sub classes.